src/gfw/common/beam/transforms/read_matching_avro_files.py

```diff
@@ -0,0 +1,180 @@
+import apache_beam as beam
+
+from apache_beam.io import fileio
+from apache_beam.io.avroio import ReadAllFromAvro
+from apache_beam.pvalue import PCollection
+
+from gfw.common.datetime import datetime_from_string, get_datetime_from_string
+
+
+logger = logging.getLogger(__name__)
+
+
+class ReadMatchingAvroFiles(beam.PTransform):
+    """A generic PTransform to filter and read Avro files from any Beam-supported filesystem.
+
+    This transform's primary function is to intelligently filter filenames
+    based on a time range. It works by:
+
+    1. **Generating Date-based Patterns**: It first generates a list of file
+       patterns for each day within the specified `start_dt` and `end_dt`. This
+       efficiently prunes the search space for large, time-partitioned datasets.
+
+    2. **Precise Datetime Filtering**: After matching the daily patterns, it
+       applies a second, more precise filter to ensure that only files with a
+       timestamp strictly within the `start_dt` and `end_dt` are processed.
+
+    This PTransform is a generic and reusable component for any data pipeline
+    that needs to perform historical data backfills on time-partitioned Avro files.
+
+    Args:
+        base_path:
+            The base path to the Avro files. This can be a local directory
+            (e.g., '/path/to/data'), a GCS bucket (e.g., 'gs://my-bucket/'),
+            or any other Beam-supported filesystem path. The transform will
+            append a date-based wildcard pattern to this path.
+
+        start_dt:
+            The start datetime of the range, in ISO format (e.g., 'YYYY-MM-DDTHH:MM:SS').
+
+        end_dt:
+            The end datetime of the range, in ISO format (e.g., 'YYYY-MM-DDTHH:MM:SS').
+            Datetimes equal to this value are considered outside the range.
+
+        path_template:
+            Path template pointing to the folder containing the avro files.
+            Since it is assumed that the data is stored in date-partitioned folders,
+            it must include a 'date' placeholder (e.g., "nmea-{date}").
+
+        decode:
+            Whether to decode the data from bytes to string.
+            Default is True.
+
+        decode_method:
+            The method used to decode the message data.
+            Supported methods include standard encodings like "utf-8", "ascii", etc.
+            Default is "utf-8".
+
+        read_all_from_avro_kwargs:
+            Any additional keyword arguments to be passed to Beam's `ReadAllFromAvro` class.
+            Check official documentation:
+            https://beam.apache.org/releases/pydoc/2.64.0/apache_beam.io.avroio.html#apache_beam.io.avroio.ReadAllFromAvro
+
+        **kwargs:
+            Additional keyword arguments passed to base PTransform class.
+
+    Returns:
+        PCollection:
+            A PCollection of Avro records from the files within the specified datetime range.
+    """
+
+    DATETIME_REGEX = r"(\d{4}-\d{2}-\d{2}).*?(\d{2}_\d{2}_\d{2}Z)"
```

Comment:
Why defining this REGEX to display it only in a log? Wouldn't be better to import `ISOFORMAT_REGEX` from datetime instead?

Reply:
The idea was to pass this regex to the function, as parameter. Even if, in this case, the default is the same. I want to be explicit. I forgot to pass it afterwards. I will change it.